In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Valens\Documents\Python Anaconda\Instacart Basket Analysis'


In [5]:
test = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data','orders_clean.csv'), index_col = False)

In [6]:
test.shape

(3421083, 8)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'orders.csv'), index_col = False)

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'products.csv'), index_col = False)

In [6]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

# Question 2
What I can see that the max in price are a little bit excessive. In this table it shows that the maximum value is 99999,where the majority price are between 4.1 - 11.2. This is why I think this is outlier.

In [7]:
# check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [8]:
#question 3 since the above does not returning mix value which means that the data is in a single data type

In [9]:
# checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [10]:
# Question 5
# there seems to be a lot of missing value in days_since_prior_order column. To be exact there is 206209

In [11]:
# creating a subset with all NaN values
# df_ords_null = df_ords.loc[df_ords['days_since_prior_order'].isnull() == True]

In [12]:
#df_ords_null

In [13]:
# creating a subset with all NaN values
# df_ords_non_null = df_ords.loc[df_ords['days_since_prior_order'].isnull() == False]

In [14]:
# df_ords_non_null

In [15]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [16]:
#replace every null with 0
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [17]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

# Question 6 
My method of dealing with the missing values would be to put 0 days. This could be that this might be the first time the user purchase the product. 

In [18]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [19]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

# Question 7 and 8
Base on the above they don't seem to be any duplicates. If there is any duplicates 

In [1]:
# This will change all pricing above 100 to be NaN
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

NameError: name 'np' is not defined

In [49]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_clean.csv'))

In [35]:
#df_ords_null.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders with null.csv'))